### PR TITLE
MSP2_INAV_SET_AUX_RC (0x2230) — Bandwidth-efficient auxiliary RC channel updates

### LINKED ISSUE #11476: Idea: MSP2_INAV_SET_AUX_RC — Lightweight Auxiliary RC Channel Update Specification
## Concept

A bandwidth-efficient MSPv2 command for updating **non-critical auxiliary RC channels** (CH9–CH32). Designed as a lightweight complement to `MSP_SET_RAW_RC` — not a replacement.

Key properties:
1. **Arbitrary start channel** — target any channel range (CH9+) without sending preceding channels
2. **Configurable resolution** — 2, 4, 8, or 16 bits per channel
3. **Non-destructive** — value `0` means "no update" (skip), the channel's previous value is preserved
4. **Persistent** — values remain set until explicitly overwritten; no timeout, no expiry
5. **Always active** — no `USE_RX_MSP` bitmask or MSP-RC-OVERRIDE flight mode required
6. **Safety-restricted** — CH1–CH8 are rejected by design (AETR + typical arm/mode channels)
7. **No failsafe interaction** — the overlay does not affect RX signal loss detection or failsafe triggering

## Usecase

Most common RC links are limited to 12 (Crossfire) or 16 channels (limit of CRSF, GHOST or SmartPort specs on radio side) but offer native MSP message passthrough. With `MSP_SET_RAW_RC` this can theoretically be extended but this message is extremely heavy, since it is forced to send 2 bytes per channel and always starts at CH1. So it gets heavier the higher the channel we would like to add. This tanks the passthrough performance way below any usable rate, if it works at all due to the split into many transport chunks that add even more delay.

`MSP2_INAV_SET_AUX_RC` solves this with predefined channel resolutions (only as large as needed) and a specific starting channel to eliminate the main overhead. This allows users to go beyond their RC link's native channel capacity to control low priority features like lights, camera triggers, gimbal modes or flap positions, without wasting precious high-resolution/low-latency native channels.

Even adding just 2 additional high-resolution channels above CH16 is significantly more compact (14 bytes vs 36 bytes, 3 SmartPort chunks vs 6).

## Safety Rationale

CH1–CH8 typically carry:
- CH1–CH4: Primary flight controls (Aileron, Elevator, Throttle, Rudder)
- CH5–CH8: Arming, flight mode selection, and other safety-critical switches

By excluding these channels, `MSP2_SET_AUX_RC` cannot be used to:
- Take over aircraft control
- Arm/disarm the aircraft
- Change flight modes

This allows the message to operate without additional safety gates (no override flag, no special flight mode required). Any MSP source (companion computer, ESP32, telemetry passthrough, Lua script, etc.) can safely update auxiliary channels at any time.

## Message Format

**MSPv2 Code:** `0x2230` (`MSP2_INAV_SET_AUX_RC`)
**Direction:** Command (to FC), fire-and-forget (no reply payload expected)
**Recommended:** Send with `MSP_FLAG_DONT_REPLY` (flags byte = `0x01`) to save bandwidth on telemetry passthrough links
**Defined in:** msp_protocol_v2_inav.h

**Payload:**

| Offset | Size | Field | Description |
|--------|------|-------|-------------|
| 0 | 1 byte | Definition byte | Start channel + resolution (see below) |
| 1 | N bytes | Channel data | Packed channel values, sequential from start channel |

**Definition Byte Encoding:**

```
Bit 7-3:  Start channel index (0-31)
            Valid range: 8-31 (CH9-CH32)
            Values 0-7 → MSP_RESULT_ERROR (CH1-CH8 protected)
Bit 2-0:  Resolution mode
            0b000 = 2-bit   (3 valid values: 1-3)
            0b001 = 4-bit   (15 valid values: 1-15)
            0b010 = 8-bit   (255 valid values: 1-255)
            0b011 = 16-bit  (raw PWM value, 1-65535)
            0b100-0b111 = reserved → MSP_RESULT_ERROR
```

**Channel count** is derived from payload size and resolution:

| Mode | Bits/CH | Channels per byte | Channel count formula |
|------|---------|-------------------|----------------------|
| 2-bit | 2 | 4 | `(payloadSize - 1) * 4` |
| 4-bit | 4 | 2 | `(payloadSize - 1) * 2` |
| 8-bit | 8 | 1 | `payloadSize - 1` |
| 16-bit | 16 | 0.5 | `(payloadSize - 1) / 2` |

**Value Scaling (receiver side):**

| Mode | Value 0 | Valid range | Mapping to PWM (µs) |
|------|---------|-------------|----------------------|
| 2-bit | skip | 1-3 | 1→1000, 2→1500, 3→2000 |
| 4-bit | skip | 1-15 | `1000 + (val-1) * 1000 / 14` (~71µs steps) |
| 8-bit | skip | 1-255 | `1000 + (val-1) * 1000 / 254` (~3.9µs steps) |
| 16-bit | skip | 1-65535 | Direct PWM value (typically 1000-2000) |

**Sub-byte packing order (2-bit and 4-bit modes):** MSB-first — the first channel occupies the highest bits of each byte.

**Validation rules (FC side):**
- `dataSize >= 2` (minimum: 1 definition byte + 1 data byte), otherwise → `MSP_RESULT_ERROR`
- `startChannel >= 8` (CH9 minimum), otherwise → `MSP_RESULT_ERROR`
- `startChannel + channelCount <= 32`, otherwise → `MSP_RESULT_ERROR`
- Resolution mode 0b100–0b111 → `MSP_RESULT_ERROR`
- 16-bit mode: data bytes must be even, otherwise → `MSP_RESULT_ERROR`
- `channelCount > 0`, otherwise → `MSP_RESULT_ERROR`

## Interaction with Existing RC System

- **No `USE_RX_MSP` required** — does not use the MSP RX provider path
- **No MSP-RC-OVERRIDE flight mode required** — always accepted
- Updates are applied as a **post-RX overlay** in `calculateRxChannelsAndUpdateFailsafe()`:
  1. Normal RX channels are read
  2. MSP RC Override is applied (if active)
  3. **MSP AUX Overlay** is applied (CH9–CH32 only, non-zero values)
  4. Failsafe evaluation runs
- Channels not included in the message (value=0 or outside range) are **not modified** — previous values persist indefinitely
- Compatible with any primary RX protocol (SBUS, CRSF, FrSKY, etc.) — acts as an overlay for aux channels only
- **No failsafe impact** — the overlay does not feed into signal-loss detection. If the primary RX link is lost, failsafe triggers normally regardless of MSP AUX overlay state

## MSPv2 Wire Format Reference

MSPv2 frame structure (for building raw packets on sender side):

```
Byte 0:     '$'  (0x24)
Byte 1:     'X'  (0x58)
Byte 2:     '<'  (0x3C)  — direction: to FC
Byte 3:     flag (0x01 recommended = DONT_REPLY, 0x00 = request reply)
Byte 4-5:   function ID = 0x2230 (little-endian: 0x30, 0x22)
Byte 6-7:   payload size (little-endian)
Byte 8..N:  payload (definition byte + channel data)
Byte N+1:   CRC8 (DVB-S2, over bytes 3..N)
```

**Total frame size** = 9 + payload_size bytes

### CRC8 DVB-S2 Algorithm (for sender implementation)

```c
uint8_t crc8_dvb_s2(uint8_t crc, uint8_t byte) {
    crc ^= byte;
    for (int i = 0; i < 8; i++) {
        if (crc & 0x80)
            crc = (crc << 1) ^ 0xD5;
        else
            crc = crc << 1;
    }
    return crc;
}

// Usage: init crc=0, feed bytes 3 through N (flag, funcID, size, payload)
```

### Example: Set CH9-CH12 to 4 switch positions (2-bit mode)

```
Definition byte: startChannel=8 (CH9), resolution=0 (2-bit)
  = (8 << 3) | 0 = 0x40

Channel data (4 channels × 2 bits = 1 byte):
  CH9=1(1000µs) CH10=3(2000µs) CH11=2(1500µs) CH12=1(1000µs)
  = 0b01_11_10_01 = 0x79
  (packed MSB-first: first channel in highest bits)

Payload: [0x40, 0x79]  (2 bytes)

Full MSPv2 frame (with DONT_REPLY):
  24 58 3C 01 30 22 02 00 40 79 [CRC8]
  $  X  <  fl funcID  sz    payload  crc
```

### Example: Set CH16-CH23 to 8-bit precision (8 channels)

```
Definition byte: startChannel=15 (CH16), resolution=2 (8-bit)
  = (15 << 3) | 2 = 0x7A

Channel data: 8 × uint8_t values (1-255, 0=skip)
  Value mapping: pwm = 1000 + (val-1) * 1000 / 254
  For 1500µs: val = round(1 + (1500-1000) * 254 / 1000) = 128

Payload: [0x7A, v1, v2, v3, v4, v5, v6, v7, v8]  (9 bytes)
Total frame: 9 + 9 = 18 bytes (vs 52 bytes with MSP_SET_RAW_RC for CH1-23)
```

## Bandwidth Comparison (MSPv2 variant)

Protocol overhead:
- MSPv1 (MSP_SET_RAW_RC): 6 bytes
- MSPv2 (MSP2_SET_AUX_RC): 9 bytes + 1 definition byte = 10 bytes fixed

### Scenario A: 8 aux channels (CH9-CH16)
| Method | CH Data | Total | Saving |
|--------|---------|-------|--------|
| MSP_SET_RAW_RC (16×u16, CH1-16) | 32 B | 38 B | baseline |
| MSP2 AUX_RC 2-bit | 2 B | 12 B | -68% |
| MSP2 AUX_RC 4-bit | 4 B | 14 B | -63% |
| MSP2 AUX_RC 8-bit | 8 B | 18 B | -53% |
| MSP2 AUX_RC 16-bit | 16 B | 26 B | -32% |

### Scenario B: 8 channels starting at CH16 (CH16-CH23)
| Method | CH Data | Total | Saving |
|--------|---------|-------|--------|
| MSP_SET_RAW_RC (CH1-23 = 23×u16) | 46 B | 52 B | baseline |
| MSP2 AUX_RC 2-bit | 2 B | 12 B | -77% |
| MSP2 AUX_RC 4-bit | 4 B | 14 B | -73% |
| MSP2 AUX_RC 8-bit | 8 B | 18 B | -65% |
| MSP2 AUX_RC 16-bit | 16 B | 26 B | -50% |

### Scenario C: 8 channels starting at CH24 (CH24-CH31)
| Method | CH Data | Total | Saving |
|--------|---------|-------|--------|
| MSP_SET_RAW_RC (CH1-31 = 31×u16) | 62 B | 68 B | baseline |
| MSP2 AUX_RC 2-bit | 2 B | 12 B | -82% |
| MSP2 AUX_RC 4-bit | 4 B | 14 B | -79% |
| MSP2 AUX_RC 8-bit | 8 B | 18 B | -74% |
| MSP2 AUX_RC 16-bit | 16 B | 26 B | -62% |

### Scenario D: 24 switch channels (CH9-CH32), 2-bit
| Method | CH Data | Total | Saving |
|--------|---------|-------|--------|
| MSP_SET_RAW_RC (32×u16) | 64 B | 70 B | baseline |
| MSP2 AUX_RC 2-bit | 6 B | 16 B | -77% |

### PR DESCRIPTION
**Pull request by b14ckyy**
Closes #11476

## Summary

Adds a new MSPv2 command `MSP2_INAV_SET_AUX_RC` (`0x2230`) that allows lightweight updates to auxiliary RC channels (CH13–CH32) with configurable resolution. Designed for extending channel count beyond native RC link capacity via MSP passthrough (SmartPort, CRSF, FPort) without the overhead of `MSP_SET_RAW_RC`.

## Motivation

Most common RC links are limited to 12–16 channels but offer native MSP passthrough. `MSP_SET_RAW_RC` is impractical over these links because it always starts at CH1, sends 2 bytes per channel, and scaling up to higher channels causes payload fragmentation across multiple transport chunks — killing passthrough performance.

`MSP2_INAV_SET_AUX_RC` solves this with:
- **Arbitrary start channel** — target any channel range CH13+ without preceding channels
- **Configurable resolution** — 2/4/8/16 bits per channel (only as much data as needed)
- **Non-destructive skip** — value `0` = no update, previous value persists

Example savings: 8 channels above CH16 in 8-bit mode = **18 bytes** vs **52 bytes** with `MSP_SET_RAW_RC` (65% reduction).

### New feature: `MSP2_INAV_SET_AUX_RC`

**`src/main/msp/msp_protocol_v2_inav.h`**
- Define `MSP2_INAV_SET_AUX_RC 0x2230`

**`src/main/rx/rx.h` / `src/main/rx/rx.c`**
- Add `static uint16_t mspAuxOverlay[]` — persistent overlay array for CH13–CH32
- Add `rxMspAuxOverlaySet(channelIndex, value)` — setter with bounds validation (index 12–31)
- Apply overlay in `calculateRxChannelsAndUpdateFailsafe()` — after MSP RC Override, before failsafe evaluation

**`src/main/rx/msp.c` / `src/main/rx/msp.h`**
- Track last `MSP_SET_RAW_RC` channel count via `rxMspGetLastChannelCount()`
- Used to prevent AUX_RC from overlapping with MSP RX provider channels

**`src/main/fc/fc_msp.c`**
- Add `case MSP2_INAV_SET_AUX_RC:` handler with:
  - Definition byte parsing (start channel + resolution mode)
  - CH1–CH12 rejection (safety guard)
  - Multi-resolution decoding: 2-bit (3 positions), 4-bit (~71µs steps), 8-bit (~3.9µs steps), 16-bit (raw PWM, clamped to 750–2250µs)
  - Sub-byte unpacking (MSB-first for 2-bit/4-bit modes)
  - Payload size validation (2–49 bytes, rejects oversized payloads to prevent uint8_t overflow)
  - Comprehensive validation (bounds, even byte count for 16-bit, reserved modes)

**`docs/development/msp/msp_messages.json`**
- Add MSP2_INAV_SET_AUX_RC entry with full field documentation

### Included bugfixes (from separate PRs, merged into base branch):
- #11478 — `initSharedMsp()` response buffer initialization fix
- #11481 — `MSP_FLAG_DONT_REPLY` handling fix for telemetry passthrough

## Safety Design

CH1–CH12 are **rejected by design** (`MSP_RESULT_ERROR`). These channels carry:
- CH1–CH4: AETR (primary flight controls)
- CH5–CH12: Arming, flight modes, and safety-critical switches. All modern RC protocols provide at least 12 channels natively — there is no valid reason to control these via AUX_RC.

This means `MSP2_SET_AUX_RC` **should not** arm/disarm, change flight modes, or take over aircraft control. No `USE_RX_MSP`, no MSP-RC-OVERRIDE flight mode, no additional safety gates required — any MSP source can safely update aux channels at any time.

### Overlay priority & safeguards

The overlay applies as the **last writer** in `calculateRxChannelsAndUpdateFailsafe()`:
1. Normal RX channels are read (SBUS/CRSF/FPort/etc.)
2. MSP RC Override is applied (if active)
3. **MSP AUX Overlay** is applied (CH13–CH32 only, non-zero values)
4. Failsafe evaluation runs (incoming AUX_RC channel values do not prevent failsafe trigger)

Three safeguards prevent unintended channel conflicts:

| Guard | Scope | Behavior |
|-------|-------|----------|
| **CH1–CH12 hard block** | All configurations | `MSP_RESULT_ERROR` — message rejected if `startChannel < 12` |
| **MSP RX channel count** | `RX_TYPE_MSP` only (e.g. mLRS) | Overlay starts at `max(12, mspLastChannelCount)` — channels covered by `MSP_SET_RAW_RC` are skipped silently |
| **MSP_RC_OVERRIDE bitmask** | When `BOXMSPRCOVERRIDE` mode is active | Channels in the `mspOverrideChannels` bitmask are skipped silently |

**Note:** For serial RX protocols (SBUS, CRSF, FPort, etc.), the protocol drivers report their maximum capacity (up to 34 channels) rather than actual sender channel count. The firmware cannot reliably detect which channels the sender actively uses. If AUX_RC targets a channel that the RX link also sends, **AUX_RC wins** (last writer). It is the user's responsibility to configure AUX_RC start channel above their RX link's active channel range.

### No failsafe interaction
The overlay does **not** affect RX signal loss detection or failsafe triggering. If the primary RX link is lost, failsafe triggers normally regardless of overlay state.

## Message Format (Quick Reference)

```
Definition byte [1 byte]:
  Bits 7-3: start channel (12-31 valid, 0-11 rejected)
  Bits 2-0: resolution (0=2-bit, 1=4-bit, 2=8-bit, 3=16-bit, 4-7=reserved/error)

Payload: [defByte] [channel data...]
  Min 2 bytes, max 49 bytes.
  Channel count derived from (payloadSize - 1) and resolution.
  Value 0 = skip (no update). Sub-byte packing: MSB-first.
  16-bit values clamped to 750-2250µs (prevents int16_t overflow in rcChannels).
```

Recommended: 
- Send with `MSP_FLAG_DONT_REPLY` (`flags=0x01`) on passthrough links for faster update rate for periodic updates
- Send MSP2_SET_AUX_RC only when channel values change and not periodically to reduce bandwidth (values are persistent) 

### Build verification
- [x] `MATEKF765` target builds clean (STM32F7, arm-none-eabi-gcc 13.2.1)
- [x] No new compiler warnings introduced

### Hardware test setup
- **FC:** MATEKF765 (STM32F765, INAV 9.0.1-dev)
- **RC link:** FrSky R9 FPort + SmartPort MSP passthrough / mLRS with CRSF Passthrough custom build
- **Sender:** Custom LUA script on radio sending raw MSPv2 frames via SmartPort/CRSF passthrough

### Functional tests performed
- [x] **4-bit mode** — Set CH13–CH20 with variable positions, verified channel movement in INAV
- [x] **8-bit mode** — Set 8 channels at ~3.9µs resolution, verified PWM scaling math is correct
- [x] **16-bit mode** — Set direct PWM values, verified raw pass-through with clamp to 750–2250
- [x] **Value 0 = skip** — Send 0 for specific channels in a batch, verified those channels retain their previous value
- [x] **Persistence** — Values remain across RX update cycles without timeout/expiry
- [x] **CH1–CH12 rejection** — StartChannel < 12 returns `MSP_RESULT_ERROR`, channels unaffected
- [x] **Bounds check** — `startChannel + channelCount > 32` returns `MSP_RESULT_ERROR`
- [x] **Reserved resolution modes** — Modes 4–7 return `MSP_RESULT_ERROR`
- [x] **16-bit odd byte** — Odd data byte count in 16-bit mode returns `MSP_RESULT_ERROR`
- [x] **Minimum payload** — `dataSize < 2` returns `MSP_RESULT_ERROR`
- [x] **Oversized payload** — `dataSize > 49` returns `MSP_RESULT_ERROR`
- [x] **DONT_REPLY flag** — With `flags=0x01`, no response frame sent on SmartPort (saves bandwidth)
- [x] **Multi-frame passthrough** — Larger payloads split across SmartPort chunks reassemble correctly
- [x] **Overlay ordering** — Values applied after MSP RC Override but before failsafe
- [x] **No failsafe impact** — Disconnecting primary RX triggers failsafe normally; overlay values do not prevent failsafe
- [x] **Coexistence with normal RX** — Aux overlay works alongside FPort primary RX without interference
- [x] **AUX_RC on active RX channels** — Verified that AUX_RC persistently overrides if targeting channels within RX link range (by design, documented)
- [x] **MSP_RC_OVERRIDE guard** — Channels in override bitmask are correctly skipped when mode is active

### Related PRs
- #11478 — `initSharedMsp()` fix (merged into base branch)
- #11481 — `MSP_FLAG_DONT_REPLY` handling fix (merged into base branch)

### CODE REVIEW

File: src/main/rx/rx.c [hunk 1]
```diff
@@ -512,6 +515,13 @@ bool calculateRxChannelsAndUpdateFailsafe(timeUs_t currentTimeUs)
     }
 #endif
 
+    // Apply MSP aux channel overlay (CH9-CH32)
+    for (int i = 8; i < 32; i++) {
+        if (mspAuxOverlay[i] > 0) {
+            rcChannels[i].data = mspAuxOverlay[i];
+        }
```
[the remark this answers is not in the document]
Comment (b14ckyy) [reply src/main/rx/rx.c]: fixed: 
```c
uint16_t pwmValue;
                    if (bitsPerChannel == 16) {
                        pwmValue = constrain(rawValue, 750, 2250);
```

File: src/main/fc/fc_msp.c [hunk 2]
```diff
@@ -2412,6 +2412,110 @@ static mspResult_e mspFcProcessInCommand(uint16_t cmdMSP, sbuf_t *src)
...
+            }
+
+            const uint8_t defByte = sbufReadU8(src);
+            const uint8_t startChannel = defByte >> 3;          // Bits 7-3: start channel index (0-31)
+            const uint8_t resolutionMode = defByte & 0x07;      // Bits 2-0: resolution
+
+            // Safety: CH1-CH8 (index 0-7) are protected
+            if (startChannel < 8) {
+                return MSP_RESULT_ERROR;
+            }
```
[the remark this answers is not in the document]
Comment (b14ckyy) [reply src/main/fc/fc_msp.c:2436]: Arming and Mode channels are common to be <CH9. Main reason for the constraint is, to not allow Control surface controls via this message. Other use cases are the operators responsibility.

**Comment by danarrib** [comment]
This is AWESOME, I'll use it for sure on Bullet GCSS. The current RAW_RC implementation always let me unease, with this I think I'll have a more reliable way to set flight modes from outside.

I'll test this and post back the results ASAP.
Thanks.

**Comment by sensei-hacker** [comment]
> This is AWESOME, I'll use it for sure on Bullet GCSS. The current RAW_RC implementation always let me unease, with this I think I'll have a more reliable way to set flight modes from outside.
> 
> I'll test this and post back the results ASAP. Thanks.

Testing is so VERY much appreciated!
Especially tests that try to find corner cases or potential problems. A lot of times when we code, we are focused on getting it to work. That focus can mean we don't think about in what cases it won't work. Or when it should NOT do the thing.  We look at trying to get it to override like it should, and don't even think about when it shouldn't. Third-party testing is awesome for that.

**Comment by b14ckyy** [comment]
- Fixed an edge case complaint from qodo with the int_8t overflow possibility. 
- added 3 priority safeguards: 
1. block AUX overlay for channels within `rxChannelCount` so no serial RX override and channel jitter possible. No error, inputs just get ignored
2. if MSP is main serial Rx provider, check what the last (highest) channel was last sent by MSP_SET_RAW_RC and only allow overlay on channels above that.
3. skip all channels within the `mspOverrideChannels` bitmask if `MSP RC OVERRIDE` mode is active to prevent dual-control of channels via MSP as well

This ensures that AUX_RC is really just acting as a low priority auxiliary control for definitely unused channels.

## tests pending

**Comment by b14ckyy** [comment]
Slight spec change. I cannot safely determine what channel range the receiver actually sends to add a reliable safeguard and reject all channels that come through sbus or crsf. Alternative solution implemented: 
- CH1-CH12 are restricted to RC link (instead 1-8) since all modern RC systems have minimum 12 channels already
- AUX_RC overlay has priority over existing RC channels since its persistent and checked last in the chain
- overlapping AUX and RC channels are the responsibility of the user (will be properly documented) 
- MSP_SET_RAW_RC and MSP2_SET_AUX_RC still has overlap check and RAW_RC has priority
- Same for MSP-Override-Bitmask

**Comment by b14ckyy** [comment]
Docs updated. Bench tested. All functions as expected. Ready to merge. Wiki Update is prepared as well, when merged.

**Comment by sensei-hacker** [comment]
@danarrib did you get a chance to check it out?

**Comment by danarrib** [comment]
Hello @sensei-hacker , sorry for the delay.

Hey @b14ckyy, just wanted to drop by and say a big thank you for this work, `MSP2_INAV_SET_AUX_RC` is working beautifully! 🎉

I integrated it into https://github.com/danarrib/BulletGCSS, my cellular-based UAV ground control station, and it made a real difference. Previously I was using `MSP_SET_RAW_RC`, which required reading and resending the full RC channel array every 200 ms just to stay within INAV's freshness window, even when nothing had changed. I also had a dependency on the MSP RC Override flight mode being active, and a gap-finder algorithm to safely neutralize channels without accidentally triggering other modes.

With `MSP2_INAV_SET_AUX_RC`, all of that is gone. I now send only when a command state actually changes, since values persist on the FC with no timeout. It's a much cleaner design and a significant simplification of my codebase.

Really appreciate the effort you put into this, great addition to INAV! 🙌

Looking forward to start using this. Hope it gets merged soon.

If you're interested in seeing the related changes on Bullet GCSS, this is the PR: https://github.com/danarrib/BulletGCSS/pull/54

Video demonstration of testing

https://github.com/user-attachments/assets/ab774a0a-e44e-4d1f-8d35-69c36002663c
